File: packages/aitrack-lib/src/data/guards.ts
/**
 * Structural type guards shared by the two hand-rolled validators: the config
 * loader and the machine-file parser. Both read untrusted JSON off disk and
 * narrow it field by field, so they need exactly the same primitives.
 */

/** A plain JSON object — arrays and null are excluded so index access is safe. */
export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** A real number: rejects NaN and ±Infinity, which survive JSON round-trips as null. */
export function isFiniteNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value);
}

File: crates/aitrack-core/src/data/guards.rs
//! Type guards and validation helpers.

use serde_json::Value;

/// A real number: rejects NaN and ±Infinity.
pub fn is_finite_number(value: &Value) -> bool {
    value.as_f64().map(|n| n.is_finite()).unwrap_or(false)
}

/// Check if value is a non-null JSON object (not an array).
pub fn is_record(value: &Value) -> bool {
    value.is_object()
}
